src/vos/ilog.c

```diff
@@ -0,0 +1,1482 @@
+			goto done;
+		}
+
+		visibility = ilog_status_get(lctx, key.id_tx_id,
+					     DAOS_INTENT_UPDATE);
+		if (visibility == ILOG_REMOVED) {
+			type = "aborted";
+			goto abort;
+		}
+		if (opc == ILOG_OP_PERSIST) {
+			if (punch)
+				break;
+			type = "redundant";
+		} else {
+			type = "punched";
+		}
+
+		if (punch && opc == ILOG_OP_PERSIST)
+			break;
+
+		if (epr->epr_hi < key.id_epoch || epr->epr_lo > key.id_epoch)
+			break;
+abort:
+		rc = remove_from_tree(lctx, type, &key, ih);
+		if (rc != 0)
+			goto done;
+	}
+collapse:
+	rc = collapse_tree(lctx, toh, ih);
+done:
+	return rc;
+}
+
+static int
+ilog_tree_modify(struct ilog_context *lctx, const struct ilog_id *id_in,
+		 bool punch, const daos_epoch_range_t *epr, int opc)
+{
+	struct ilog_root	*root;
+	bool			*punchp;
+	struct ilog_id		*keyp;
+	struct ilog_id		 key;
+	struct ilog_id		 id = *id_in;
+	daos_handle_t		 toh = DAOS_HDL_INVAL;
+	daos_handle_t		 ih = DAOS_HDL_INVAL;
+	d_iov_t			 key_iov;
+	d_iov_t			 val_iov;
+	int			 visibility = ILOG_VISIBLE;
+	struct umem_attr	 uma;
+	int			 rc = 0;
+
+	root = lctx->ic_root;
+
+	umem_attr_get(&lctx->ic_umm, &uma);
+	rc = dbtree_open(root->lr_tree.it_root, &uma, &toh);
+	if (rc != 0) {
+		D_ERROR("Failed to open incarnation log tree: rc = %s\n",
+			d_errstr(rc));
+		goto done;
+	}
+
+	rc = dbtree_iter_prepare(toh, BTR_ITER_EMBEDDED, &ih);
+	if (rc != 0) {
+		D_ERROR("Could not prepare iterator: rc = %s\n", d_errstr(rc));
+		return rc;
+	}
+
+	for (;;) {
+		bool	is_equal;
+
+		d_iov_set(&key_iov, (struct ilog_id *)&id, sizeof(id));
+
+		rc = dbtree_iter_probe(ih, BTR_PROBE_LE, DAOS_INTENT_DEFAULT,
+				       &key_iov, NULL);
+		if (rc == -DER_NONEXIST)
+			break; /* Skip to handler */
+		if (rc != 0) {
+			D_ERROR("Could not probe iterator: rc = %s\n",
+				d_errstr(rc));
+			goto done;
+		}
+
+		d_iov_set(&key_iov, NULL, 0);
+		d_iov_set(&val_iov, NULL, 0);
+
+		rc = dbtree_iter_fetch(ih, &key_iov, &val_iov, NULL);
+		if (rc != 0) {
+			D_ERROR("Could not fetch from iterator: rc = %s\n",
+				d_errstr(rc));
+			goto done;
+		}
+
+		punchp = (bool *)val_iov.iov_buf;
+		keyp = (struct ilog_id *)key_iov.iov_buf;
+
+		visibility = ilog_status_get(lctx, keyp->id_tx_id,
+					     DAOS_INTENT_UPDATE);
+
+		if (visibility == ILOG_REMOVED) {
+			rc = remove_from_tree(lctx, "aborted", keyp, &ih);
+			if (rc != 0)
+				goto done;
+			continue;
+		}
+
+		rc = update_inplace(lctx, keyp, punchp, id_in, opc, punch,
+				    &is_equal);
+		if (rc != 0)
+			goto done;
+
+		if (is_equal) {
+			if (opc == ILOG_OP_UPDATE)
+				goto done;
+
+			rc = consolidate_tree(lctx, epr, &toh, &ih, opc,
+					      id_in, *punchp);
+
+			goto done;
+		}
+
+		if (opc != ILOG_OP_UPDATE) {
+			D_DEBUG(DB_IO, "No entry found, done\n");
+			goto done;
+		}
+
+		if (punch || visibility == ILOG_INVISIBLE || *punchp)
+			break; /* handle entry */
+		/* the new update is "covered" by a previous one */
+		goto done;
+	}
+
+	rc = ilog_tx_begin(lctx);
```

Comment:
The ilog_tx_begin() are scattered in modify/destroy/migrate/ptr_set, but related ilog_tx_end() may be not in the same caller functions. Such non-pairs usage is some confused. In addition, the caller itself may be within another PMDK transaction. That makes the things more complex. So I would suggest to make the caller to explicitly call ilog_tx_begin and ilog_tx_end() in pairs, that is more intuitive and clean.

Reply:
Indeed.   I was trying to avoid starting an unnecessary transaction.   The idea is to start one on demand and track that we've started it so when we call end, it also exists the transaction.

We could remove the tx code altogether as in VOS, we are generally in a transaction but this is a tradeoff that allows me to test it independently.